go/mysql/binlog_event_compression.go

```diff
@@ -71,8 +73,24 @@ var (
 
 	// Create a stateless reader that caches decompressors. This is used
 	// for smaller events that we want to handle entirely using in-memory
-	// buffers.
-	zstdDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
+	// buffers via DecodeAll.
+	statelessDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
+
+	// Use a pool of stateful decoders for larger payloads that we want
+	// to stream. The number of large (> zstdInMemoryDecompressorMaxSize)
+	// payloads should typically be relatively low, but there may be times
+	// where there are many of them -- and users like vstreamer may have
+	// N concurrent streams per vttablet which could lead to a lot of
+	// allocations and GC overhead so this pool allows us to handle
+	// concurrent cases better while still scaling to 0 when there's no
+	// usage.
+	statefulDecodersPool = sync.Pool{
+		New: func() any {
+			d, err := zstd.NewReader(nil, zstd.WithDecoderMaxMemory(zstdInMemoryDecompressorMaxSize))
+			log.Errorf("Error creating stateful decoder: %v", err)
```

Comment:
I saw this a couple days ago that helps propagate the real error instead of checking for nil, but what you have is probably sufficient. It doesn't seem like that new function should error out in practice
https://www.reddit.com/r/golang/comments/1dww9bi/how_to_handle_errors_on_synvpool/lbxinal